docs/source/notes/extending.rst

```diff
@@ -204,6 +204,253 @@ This is how a ``Linear`` module can be implemented::
+          HANDLED_FUNCTIONS[torch_function] = func
+          return func
+      return decorator
+
+which can be applied to the implementation of our override::
+
+  @implements(torch.mean)
+  def mean(input):
+      return float(input._value) / input._N
+
+With this change we can now use ``torch.mean`` with ``ScalarTensor``::
+
+  >>> d = ScalarTensor(5, 2)
+  >>> torch.mean(d)
+  0.4
+
+Of course ``torch.mean`` is an example of the simplest kind of function to
+override since it only takes one operand. We can use the same machinery to
+override a function that takes more than one operand, any one of which might be
+a tensor or tensor-like that defines ``__torch_function__``, for example for
+:func:`torch.add`::
+
+  def ensure_tensor(data):
+      if isinstance(data, ScalarTensor):
+          return data.tensor()
+      return torch.as_tensor(data)
+
+  @implements(torch.add)
+  def add(input, other):
+     try:
+         if input._N == other._N:
+             return ScalarTensor(input._N, input._value + other._value)
+         else:
+             raise ValueError("Shape mismatch!")
+     except AttributeError:
+         return torch.add(ensure_tensor(input), ensure_tensor(other))
+
+This version has a fast path for when both operands are ``ScalarTensor``
+instances and also a slower path whcih degrades to converting the data to
+tensors when either operand is not a ``ScalarTensor``. That makes the override
+function correctly when either operand is a ``ScalarTensor`` or a regular
+:class:`Tensor`::
+
+  >>> s = ScalarTensor(2, 2)
+  >>> torch.add(s, s)
+  DiagonalTensor(N=2, value=4)
+  >>> t = torch.tensor([[1, 1,], [1, 1]])
+  >>> torch.add(s, t)
+  tensor([[3., 1.],
+          [1., 3.]])
+
+Note that our implementation of ``add`` does not take ``alpha`` or ``out`` as
+keyword arguments like :func:`torch.add` does::
+
+  >>> torch.add(s, s, alpha=2)
+  TypeError: add() got an unexpected keyword argument 'alpha'
+
+For speed and flexibility the ``__torch_function__`` dispatch mechanism does not
+check that the signature of an override function matches the signature of the
+function being overrided in the :mod:`torch` API. For some applications ignoring
+optional arguments would be fine but to ensure full compatibility with
+:class:`Tensor`, user implementations of torch API functions should take care to
+exactly emulate the API of the function that is being overrided.
+
+Functions in the :mod:`torch` API that do not have explicit overrides will
+return ``NotImplemented`` from ``__torch_function__``. If all operands with
+``__torch_function__`` defined on them return ``NotImplemented``, PyTorch will
+raise a ``TypeError``. This means that most of the time operations that do not
+have explicit overrides for a type will raise a ``TypeError`` when an instance
+of such a type is passed::
+
+  >>> torch.mul(s, 3)
+  TypeError: no implementation found for 'torch.mul' on types that
+  implement __torch_function__: [ScalarTensor]
```

Comment:
There's a critical part missing here I think: the consequences this has. For the API to work with the current example that raises `NotImplemented` unless the function is handled, it in effect means that users of this functionality should implement _all_ functions (or all they care about). That's a lot of work, so better to be clear about it.

An alternative solution here is for functions that aren't handled, coercing to plain `Tensor` and calling the `torch.<func>` function again with that coerced input. That way the extra metadata/methods/etc get lost (which could warrant a warning), but in many cases is more practical than a `TypeError`.
Dask has this behavior for `__array_function__`, it took a little while to get right, but it's probably good to steal that and add it to the docs here.

Reply:
That’s what I was going for with `MetadataTensor` below. Do you think a sentence acknowledging the limitation you brought up and referencing that example as one that does something along those lines here would be sufficient or do you think I should do something more in-depth?

Reply:
I don't think that `MetadataTensor` is very related. It adds zero function overrides, and the `HANDLED_FUNCTIONS` pattern to add individual overrides isn't used there.

I think it'd be good to have an extension to this section that instead of using `NotImplemented`, does something like the `handled_nonmatching_names` from https://github.com/dask/dask/pull/5043

Reply:
Or even just describe it in words if the implementation would be too long. The logic is pretty simple.

Reply:
I ended up expanding the `ScalarTensor` example a bit then referenced `MetadataTensor` as well. Thanks for the review, this is a great point that's worth going into more detail on.

Reply:
Looks good, thanks Nathan.

Reply:
Looks like this is already being landed, so you may have to resubmit that last commit in a few days.